flink-end-to-end-tests/src/main/java/org/apache/flink/streaming/tests/StickyAllocationAndLocalRecoveryTestJob.java

```diff
@@ -0,0 +1,451 @@
+			env.getCheckpointConfig().enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
+		}
+
+		String stateBackend = pt.get("stateBackend", "file");
+		String checkpointDir = pt.getRequired("checkpointDir");
+
+		boolean killJvmOnFail = pt.getBoolean("killJvmOnFail", false);
+
+		if ("file".equals(stateBackend)) {
+			boolean asyncCheckpoints = pt.getBoolean("asyncCheckpoints", false);
+			env.setStateBackend(new FsStateBackend(checkpointDir, asyncCheckpoints));
+		} else if ("rocks".equals(stateBackend)) {
+			boolean incrementalCheckpoints = pt.getBoolean("incrementalCheckpoints", false);
+			env.setStateBackend(new RocksDBStateBackend(checkpointDir, incrementalCheckpoints));
+		} else {
+			throw new IllegalArgumentException("Unknown backend: " + stateBackend);
+		}
+
+		// make parameters available in the web interface
+		env.getConfig().setGlobalJobParameters(pt);
+
+		// delay to throttle down the production of the source
+		long delay = pt.has("delay") ? pt.getLong("delay") : 0L;
+
+		// the maximum number of attempts, before the job finishes with success
+		int maxAttempts = pt.has("maxAttempts") ? pt.getInt("maxAttempts") : 3;
+
+		// size of one artificial value
+		int valueSize = pt.has("valueSize") ? pt.getInt("valueSize") : 10;
+
+		env.addSource(new RandomLongSource(maxAttempts, delay))
+			.keyBy((KeySelector<Long, Long>) aLong -> aLong)
+			.flatMap(new StateCreatingFlatMap(valueSize, killJvmOnFail))
+			.map((MapFunction<String, String>) value -> value)
+			.addSink(new PrintSinkFunction<>());
+
+		env.execute("Sticky Allocation And Local Recovery Test");
+	}
+
+	/**
+	 * Source function that produces a long sequence.
+	 */
+	private static final class RandomLongSource extends RichSourceFunction<Long> implements CheckpointedFunction {
+
+		private static final long serialVersionUID = 1L;
+
+		/**
+		 * Generator delay between two events.
+		 */
+		final long delay;
+
+		/**
+		 * Maximum restarts before shutting down this source.
+		 */
+		final int maxAttempts;
+
+		/**
+		 * State that holds the current key for recovery.
+		 */
+		ListState<Long> sourceCurrentKeyState;
+
+		/**
+		 * Generator's current key.
+		 */
+		long currentKey;
+
+		/**
+		 * Generator runs while this is true.
+		 */
+		volatile boolean running;
+
+		RandomLongSource(int maxAttempts, long delay) {
+			this.delay = delay;
+			this.maxAttempts = maxAttempts;
+			this.running = true;
+		}
+
+		@Override
+		public void run(SourceContext<Long> sourceContext) throws Exception {
+
+			int numberOfParallelSubtasks = getRuntimeContext().getNumberOfParallelSubtasks();
+			int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+
+			// the source emits one final event and shuts down once we have reached max attempts.
+			if (getRuntimeContext().getAttemptNumber() > maxAttempts) {
+				sourceContext.collect(Long.MAX_VALUE - subtaskIdx);
```

Comment:
element emission should always happen under the checkpoint lock

Reply:
👍